crates/dht/src/dht/snapshot.rs

```diff
@@ -0,0 +1,763 @@
+    /// for that sector.
+    DiscSectorDetails {
+        /// Similar to the `disc_sector_top_hashes` except the full time slice hashes are not
+        /// combined.
+        disc_sector_hashes: HashMap<u32, HashMap<u64, bytes::Bytes>>,
+        /// The end timestamp of the most recent full time slice.
+        disc_boundary: Timestamp,
+    },
+    /// A snapshot to be used when there is a [DhtSnapshot::Minimal] mismatch in the ring top
+    /// hashes.
+    RingSectorDetails {
+        /// Similar to the `ring_top_hashes` except the sector hashes are not combined.
+        ring_sector_hashes: HashMap<u32, HashMap<u32, bytes::Bytes>>,
+        /// The end timestamp of the most recent full time slice.
+        disc_boundary: Timestamp,
+    },
+}
+
+impl DhtSnapshot {
+    /// Compare two snapshots to determine how they differ.
+    ///
+    /// Produces a [SnapshotDiff] that describes the differences between the two snapshots.
+    /// This should not be use directly, please see [crate::dht::Dht::handle_snapshot].
+    pub fn compare(&self, other: &Self) -> SnapshotDiff {
+        // Check if they match exactly, before doing further work to check how they differ.
+        if self == other {
+            return SnapshotDiff::Identical;
+        }
+
+        match (self, other) {
+            (
+                DhtSnapshot::Minimal {
+                    disc_top_hash: our_disc_top_hash,
+                    disc_boundary: our_disc_boundary,
+                    ring_top_hashes: our_ring_top_hashes,
+                },
+                DhtSnapshot::Minimal {
+                    disc_top_hash: other_disc_top_hash,
+                    disc_boundary: other_disc_boundary,
+                    ring_top_hashes: other_ring_top_hashes,
+                },
+            ) => {
+                // If the historical time boundary doesn't match, we can't compare.
+                // This won't happen very often so it's okay to just fail this match.
+                if our_disc_boundary != other_disc_boundary {
+                    return SnapshotDiff::CannotCompare;
+                }
+
+                // If the disc hash mismatches, then there is a historical mismatch.
+                // This shouldn't be common, but we sync forwards through time, so if we
+                // find a historical mismatch then focus on fixing that first.
+                if our_disc_top_hash != other_disc_top_hash {
+                    return SnapshotDiff::DiscMismatch;
+                }
+
+                // This is more common, it can happen if we're close to a UNIT_TIME boundary
+                // and there is a small clock difference or just one node calculated this snapshot
+                // before the other did. Still, we'll have to wait until we next compare to
+                // compare our DHT state.
+                if our_ring_top_hashes.len() != other_ring_top_hashes.len() {
+                    return SnapshotDiff::CannotCompare;
+                }
```

Comment:
I assume, if these are off by one because one of the nodes moved forward, we cannot just ignore that extra, because the remaining rings may be re-sized so the hashes won't match?

What is the effect of returning a CannotCompare do you imagine? Will the node's just retry after a bit when they should again be in the same time-state?

Reply:
Exactly that, if the times are off by enough for the rings to change then we can't work with indices safely. I believe the representation of partial time slices is unique so if you pin the end of the full time slices and check the number of partial slices, I think you can be sure you're comparing the same time slices.

Yes I think it has to terminate the sync process early. So we'll still do gossip of agents and "what's new since we last gossiped" and then this part would have to fail if the model doesn't match.

If it gets updated mid round that would kill the round so at least for rounds we've initiated, I think we should avoid doing that during the round. Right before we initiate would be a good time. Then rounds initiated by others might then fail but the model only needs updating every UNIT_TIME which is 15 minutes, so even if we only update every 5 minutes, there has to be a pretty significant clock skew for a mismatch to happen I think